### PR TITLE
[Feature] -New caffe transformers

### PR DESCRIPTION
**Pull request by wzhongyuan**
## What changes were proposed in this pull request?

Adding new image frame based transformers

## How was this patch tested?

unit test

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/augmentation/ChannelScaledNormalizer.scala [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+import com.intel.analytics.bigdl.dataset.image.LabeledBGRImage
+import com.intel.analytics.bigdl.dataset.{LocalDataSet, Transformer}
+import com.intel.analytics.bigdl.transform.vision.image.{FeatureTransformer, ImageFeature}
+import com.intel.analytics.bigdl.transform.vision.image.opencv.OpenCVMat
+import org.apache.log4j.Logger
+
+import scala.collection.Iterator
+
+object ChannelScaledNormalizer {
+  val logger = Logger.getLogger(getClass)
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/augmentation/ChannelScaledNormalizer.scala]: private?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/augmentation/ChannelScaledNormalizer.scala [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+  def apply(meanR: Int, meanG: Int, meanB: Int, scale: Double): ChannelScaledNormalizer = {
+    new ChannelScaledNormalizer(meanR, meanG, meanB, scale)
+  }
+}
+
+class ChannelScaledNormalizer(meanR: Int, meanG: Int, meanB: Int, scale: Double)
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/augmentation/ChannelScaledNormalizer.scala]: Java doc?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/augmentation/RandomAlterAspect.scala [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+class RandomAlterAspect(min_area_ratio: Float = 0.08f,
+                           max_area_ratio: Int = 1,
+                           min_aspect_ratio_change: Float = 0.75f,
+                           interp_mode: String = "CUBIC",
+                           cropLength: Int = 224)
+  extends FeatureTransformer {
+
+  import com.intel.analytics.bigdl.utils.RandomGenerator.RNG
+
+  def randRatio(min: Float, max: Float): Float = {
```
[the remark this answers is not in the document]
Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/augmentation/RandomAlterAspect.scala]: it's specific to resnet-50 so I make it as argument to make it more common with default value

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/augmentation/RandomAlterAspect.scala [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+            min_aspect_ratio_change: Float = 0.75f,
+            interp_mode: String = "CUBIC",
+            cropLength: Int = 224): RandomAlterAspect = {
+    OpenCV.isOpenCVLoaded
+    new RandomAlterAspect(min_area_ratio, max_area_ratio,
+      min_aspect_ratio_change, interp_mode, cropLength)
+  }
+}
+
+class RandomAlterAspect(min_area_ratio: Float = 0.08f,
```
[the remark this answers is not in the document]
Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/augmentation/RandomAlterAspect.scala]: Added for all 4 transformers

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/augmentation/RandomAlterAspect.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (yiheng) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/augmentation/RandomAlterAspect.scala]: is it a part of the API for user? if so we need to provide a java doc

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/augmentation/RandomCropper.scala [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+    if (openCVMat.`type`() != CvType.CV_32FC3) {
+      openCVMat.convertTo(openCVMat, CvType.CV_32FC3)
+    }
+
+    val tmp = new Array[Float](openCVMat.height() * openCVMat.width() * openCVMat.channels())
+
+    openCVMat.get(0, 0, tmp)
+
+    val buffer = new Array[Float](cropWidth * cropHeight * openCVMat.channels)
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/augmentation/RandomCropper.scala]: reuse the buffer?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/augmentation/RandomCropper.scala [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+  import com.intel.analytics.bigdl.utils.RandomGenerator.RNG
+
+  override protected def transformMat(feature: ImageFeature): Unit = {
+    val openCVMat = feature.opencvMat()
+
+    if (openCVMat.`type`() != CvType.CV_32FC3) {
+      openCVMat.convertTo(openCVMat, CvType.CV_32FC3)
+    }
+
+    val tmp = new Array[Float](openCVMat.height() * openCVMat.width() * openCVMat.channels())
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/augmentation/RandomCropper.scala]: reuse the buffer?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/augmentation/RandomCropper.scala [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+
+    openCVMat.get(0, 0, tmp)
+
+    val buffer = new Array[Float](cropWidth * cropHeight * openCVMat.channels)
+    val height = openCVMat.size().height.toInt
+    val width = openCVMat.size().width.toInt
+
+    val bytes = new Array[Float](height * width * 3)
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/augmentation/RandomCropper.scala]: reuse the buffer?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/augmentation/RandomCropper.scala [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+      openCVMat.release()
+    }
+    if (mate.`type`() != CvType.CV_32FC3) {
+      mate.convertTo(mate, CvType.CV_32FC3)
+    }
+    mate.put(0, 0, buffer)
+    feature(ImageFeature.mat) = mate
+  }
+
+  def cropper(source: Array[Float], target: Array[Float], srcSize: Array[Int],
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/augmentation/RandomCropper.scala]: private?

**Comment by yiheng** [comment]
will the opencv call native multithread?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/augmentation/ChannelScaledNormalizer.scala [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+  extends FeatureTransformer {
+
+  override protected def transformMat(feature: ImageFeature): Unit = {
+    val mat = feature.opencvMat()
+    val toFloats = OpenCVMat.toFloatPixels(mat)
+    val content = toFloats._1
+    require(content.length % 3 == 0)
```
[the remark this answers is not in the document]
Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/augmentation/ChannelScaledNormalizer.scala]: done

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/augmentation/RandomAlterAspect.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/augmentation/RandomAlterAspect.scala]: no, i've made it private, i thought you mean the crop length

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/augmentation/RandomCropper.scala (hunk 5 above)
Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/augmentation/RandomCropper.scala]: yes. made it global

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/augmentation/RandomAlterAspect.scala [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+      attempt += 1
+    }
+    if (attempt < 20) {
+      val (new_h, new_w) = resizeImagePerShorterSize(feature.opencvMat(), cropLength)
+      Imgproc.resize(feature.opencvMat(),
+        feature.opencvMat(), new Size(cropLength, cropLength), 0, 0, 2)
+    }
+  }
+
+  def resizeImagePerShorterSize(img: Mat, shorter_size: Int) : (Int, Int) = {
```
[the remark this answers is not in the document]
Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/augmentation/RandomAlterAspect.scala]: done

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/augmentation/RandomCropper.scala (hunk 6 above)
Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/augmentation/RandomCropper.scala]: this is image specific, cannot make it global

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/augmentation/RandomCropper.scala (hunk 7 above)
Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/augmentation/RandomCropper.scala]: it's image specific

**Comment by wzhongyuan** [comment]
@yiheng  I don't think it's multi thread, the java interface is one-to-one mapping to native

**Comment by wzhongyuan** [comment]
PR validation passed http://mlt-bddl01.sh.intel.com:8080/job/BigDL-PRVN-lite/25/

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/augmentation/ChannelScaledNormalizer.scala [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+    var i = 0
+    val frameLength = content.length / 3
+    val height = toFloats._2
+    val width = toFloats._3
+    val bufferContent = new Array[Float](width * height * 3)
```
[the remark this answers is not in the document]
Comment (yiheng) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/augmentation/ChannelScaledNormalizer.scala]: only recreate the buffer when the image need a larger buffer?

**Comment by wzhongyuan** [comment]
@yiheng   As discussed offline, removed the unused code, others are left as they are as they are args for OpenCV

**Comment by yiheng** [comment]
after jenkins pass, let's merge it.

**Comment by wzhongyuan** [comment]
PR validation passed http://mlt-bddl01.sh.intel.com:8080/job/BigDL-PRVN-lite/28/

**Pull request merged**